examples/fluids/src/misc.c

```diff
@@ -132,32 +132,126 @@ PetscErrorCode RegressionTests_NS(AppCtx app_ctx, Vec Q) {
   PetscFunctionReturn(0);
 }
 
-// Get error for problems with exact solutions
-PetscErrorCode GetError_NS(CeedData ceed_data, DM dm, User user, Vec Q, PetscScalar final_time) {
-  PetscInt  loc_nodes;
-  Vec       Q_exact, Q_exact_loc;
-  PetscReal rel_error, norm_error, norm_exact;
-  PetscFunctionBegin;
+PetscErrorCode ConvertStateVar_NS(DM dm, CeedData ceed_data, User user, Vec Qsource_loc, Vec Qtarget_loc, Vec Qtarget) {
+  PetscFunctionBeginUser;
+  // CEED Restriction
+  CeedVector q_source_ceed, q_target_ceed;
+  CeedElemRestrictionCreateVector(ceed_data->elem_restr_q, &q_source_ceed, NULL);
+  CeedElemRestrictionCreateVector(ceed_data->elem_restr_q, &q_target_ceed, NULL);
+
+  // Place PETSc vectors in CEED vectors
+  PetscScalar *q_source, *q_target;
+  PetscMemType q_source_mem_type, q_target_mem_type;
+  PetscCall(VecGetArrayReadAndMemType(Qsource_loc, (const PetscScalar **)&q_source, &q_source_mem_type));
+  CeedVectorSetArray(q_source_ceed, MemTypeP2C(q_source_mem_type), CEED_USE_POINTER, q_source);
+  PetscCall(VecZeroEntries(Qtarget_loc));
+  PetscCall(VecGetArrayAndMemType(Qtarget_loc, (PetscScalar **)&q_target, &q_target_mem_type));
+  CeedVectorSetArray(q_target_ceed, MemTypeP2C(q_target_mem_type), CEED_USE_POINTER, q_target);
+
+  // Apply CEED Operator
+  CeedOperatorApply(ceed_data->op_to_prim, q_source_ceed, q_target_ceed, CEED_REQUEST_IMMEDIATE);
+
+  // Restore vectors
+  CeedVectorTakeArray(q_source_ceed, MemTypeP2C(q_source_mem_type), NULL);
+  PetscCall(VecRestoreArrayReadAndMemType(Qsource_loc, (const PetscScalar **)&q_source));
+  CeedVectorTakeArray(q_target_ceed, MemTypeP2C(q_target_mem_type), NULL);
+  PetscCall(VecRestoreArrayReadAndMemType(Qtarget_loc, (const PetscScalar **)&q_target));
+
+  // Local-to-Global
+  PetscCall(VecZeroEntries(Qtarget));
+  PetscCall(DMLocalToGlobal(dm, Qtarget_loc, ADD_VALUES, Qtarget));
+
+  // Cleanup
+  CeedVectorDestroy(&q_source_ceed);
+  CeedVectorDestroy(&q_target_ceed);
+
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode ComputeErrorComponent_NS(Vec Q_true_loc, Vec Q_soln_loc, NormType norm_type, MPI_Op norm_reduce, PetscReal norm_true[5],
+                                        PetscReal norm_error[5], PetscReal rel_error[5]) {
+  // Get the norm of each component
+  PetscCall(VecStrideNormAll(Q_true_loc, norm_type, norm_true));
+  PetscCallMPI(MPI_Allreduce(MPI_IN_PLACE, norm_true, 5, MPIU_REAL, norm_reduce, PETSC_COMM_WORLD));
+  // Compute the error of each component
+  PetscCall(VecAXPY(Q_soln_loc, -1.0, Q_true_loc));
+  PetscCall(VecStrideNormAll(Q_soln_loc, norm_type, norm_error));
+  PetscCallMPI(MPI_Allreduce(MPI_IN_PLACE, norm_error, 5, MPIU_REAL, norm_reduce, PETSC_COMM_WORLD));
+  for (int i = 0; i < 5; i++) {
+    rel_error[i] = norm_error[i] / norm_true[i];
+  }
```

Comment:
I think this can only be correct as written when both norms are `MAX`. So maybe don't bother passing them in. If we want $L^2$ norms, we'd have to do a little work.

Reply:
Done!